Sources/RSocketCore/Streams/ThreadSafeStreamAdapter.swift

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2015-present the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import NIO
+
+internal protocol StreamAdapterDelegate: AnyObject {
+    func send(frame: Frame)
+}
+
+/// `ThreadSafeStreamAdapter` converts all stream events (e.g. onNext, onCancel, etc,) into frames with the given `id` and forwards them to `delegate` on `eventLoop`
+internal final class ThreadSafeStreamAdapter {
+    internal var id: StreamID!
+    private let eventLoop: EventLoop
+    internal weak var delegate: StreamAdapterDelegate?
+    
+    internal init(id: StreamID? = nil, eventLoop: EventLoop, delegate: StreamAdapterDelegate? = nil) {
+        self.id = id
+        self.eventLoop = eventLoop
+        self.delegate = delegate
+    }
+}
+
+extension ThreadSafeStreamAdapter: UnidirectionalStream {
+    private func send<Body>(_ body: Body) where Body: FrameBodyBoundToStream {
+        eventLoop.execute { [self] in
+            self.delegate?.send(frame: body.asFrame(withStreamId: self.id))
+        }
```

Comment:
can we do a check if it is already on the eventloop before scheduling something?

```suggestion
        if eventLoop.inEventLoop {
            self.delegate?.send(frame: body.asFrame(withStreamId: self.id))
        } else {
            eventLoop.execute { [self] in
                self.delegate?.send(frame: body.asFrame(withStreamId: self.id))
            }
        } 
```

Reply:
I guess the block related to streamId generation has to have the same check as in the TreadSafeStreamAdapter. Having both of them check if it is already on the event loop will guarantee serial message sending and ensure that steamId generation happens before the first message being sent. Otherwise, `eventLoop.execute { [self] in` happens always before any other execution and happens before the adapter is given to the user hands